addons/mrp/models/mrp_production.py

```diff
@@ -2614,39 +2613,38 @@ def _get_ratio_between_mo_and_bom_quantities(self, bom):
 
     def _check_sn_uniqueness(self):
         """ Alert the user if the serial number as already been consumed/produced """
-        if self.product_tracking == 'serial' and self.lot_producing_id:
-            if self._is_finished_sn_already_produced(self.lot_producing_id):
-                raise UserError(_('This serial number for product %s has already been produced', self.product_id.name))
-
-        for move in self.move_finished_ids:
-            if move.has_tracking != 'serial' or move.product_id == self.product_id:
-                continue
-            for move_line in move.move_line_ids:
-                if float_is_zero(move_line.quantity, precision_rounding=move_line.product_uom_id.rounding):
-                    continue
-                if self._is_finished_sn_already_produced(move_line.lot_id, excluded_sml=move_line):
-                    raise UserError(_('The serial number %(number)s used for byproduct %(product_name)s has already been produced',
-                                      number=move_line.lot_id.name, product_name=move_line.product_id.name))
-
+        serial_tracked_productions = self.filtered(lambda production: production.product_tracking == 'serial' and self.lot_producing_id)
+        duplicate_producing_lot = serial_tracked_productions._is_finished_sn_already_produced_batch(serial_tracked_productions.mapped('lot_producing_id'))
+        if duplicate_producing_lot:
+            raise UserError(_('This serial number for product %s has already been produced', duplicate_producing_lot.product_id.name))
+
+        serial_tracked_moves = self.mapped('move_finished_ids').filtered(lambda move: move.has_tracking == 'serial' and move.product_id != self.product_id)
+        to_check_move_lines = serial_tracked_moves.mapped('move_line_ids').filtered(lambda move_line: not float_is_zero(move_line.quantity, precision_rounding=move_line.product_uom_id.rounding))
```

Comment:
I think you can do this in a single filter.
Basically, `move_finished_ids` that have the same product as the MO as the main finished product (not the byproducts), and this should almost always have a single move line.
So I don't think you're skipping much by doing this in two separate instances 🤔